File: benchmarks/pandas/bench_advance_date_fn.py
"""
Benchmark: pandas DateOffset arithmetic — date frequency parsing and advancement.
Mirrors tsb advanceDate / parseFreq.
Outputs JSON: {"function": "advance_date_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd

WARMUP = 5
ITERATIONS = 1000

d = pd.Timestamp("2023-06-15")
offsets = [
    pd.DateOffset(days=1),
    pd.DateOffset(days=3),
    pd.offsets.BDay(1),
    pd.offsets.Week(1),
    pd.offsets.MonthBegin(1),
    pd.offsets.MonthEnd(1),
    pd.DateOffset(hours=1),
    pd.DateOffset(hours=2),
    pd.DateOffset(minutes=1),
    pd.offsets.YearBegin(1),
]

for _ in range(WARMUP):
    for off in offsets:
        d + off
    pd.Timestamp("2023-01-01")
    pd.Timestamp(1672531200000, unit="ms")

t0 = time.perf_counter()
for _ in range(ITERATIONS):
    for off in offsets:
        d + off
    pd.Timestamp("2023-01-01")
    pd.Timestamp(1672531200000, unit="ms")
total = (time.perf_counter() - t0) * 1000

print(json.dumps({
    "function": "advance_date_fn",
    "mean_ms": round(total / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total, 3),
}))

File: benchmarks/tsb/bench_advance_date_fn.ts
/**
 * Benchmark: advanceDate / parseFreq — date frequency utilities.
 * Mirrors pandas DateOffset arithmetic.
 * Outputs JSON: {"function": "advance_date_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { advanceDate, parseFreq, toDateInput } from "../../src/index.ts";

const WARMUP = 5;
const ITERATIONS = 1000;

const d = new Date("2023-06-15");
const freqSpecs = ["D", "3D", "B", "W", "MS", "ME", "h", "2h", "min", "YS"];

for (let i = 0; i < WARMUP; i++) {
  for (const f of freqSpecs) {
    const pf = parseFreq(f);
    advanceDate(d, pf);
  }
  toDateInput("2023-01-01");
  toDateInput(1672531200000);
}

const t0 = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  for (const f of freqSpecs) {
    const pf = parseFreq(f);
    advanceDate(d, pf);
  }
  toDateInput("2023-01-01");
  toDateInput(1672531200000);
}
const total = performance.now() - t0;

console.log(
  JSON.stringify({
    function: "advance_date_fn",
    mean_ms: Math.round((total / ITERATIONS) * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(total * 1000) / 1000,
  }),
);
